syntax = "proto3";

package discord_action;

// Discord Action Service
service DiscordAction {
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
  rpc SendEmbed(SendEmbedRequest) returns (SendEmbedResponse);
  rpc AddReaction(AddReactionRequest) returns (AddReactionResponse);
  rpc ManageRole(ManageRoleRequest) returns (ManageRoleResponse);
  rpc CreateWebhook(CreateWebhookRequest) returns (CreateWebhookResponse);
  rpc SendWebhook(SendWebhookRequest) returns (SendWebhookResponse);
  rpc DeleteMessage(DeleteMessageRequest) returns (DeleteMessageResponse);
  rpc EditMessage(EditMessageRequest) returns (EditMessageResponse);
  rpc KickUser(KickUserRequest) returns (KickUserResponse);
  rpc BanUser(BanUserRequest) returns (BanUserResponse);
  rpc TimeoutUser(TimeoutUserRequest) returns (TimeoutUserResponse);
}

// Message structures
message SendMessageRequest {
  string channel_id = 1;
  string content = 2;
  optional EmbedData embed = 3;
  string token = 4;  // JWT token
}

message SendMessageResponse {
  bool success = 1;
  string message_id = 2;
  string error = 3;
}

message EmbedData {
  string title = 1;
  string description = 2;
  string color = 3;
  string url = 4;
  string thumbnail_url = 5;
  string image_url = 6;
  string footer_text = 7;
  string footer_icon_url = 8;
  repeated EmbedField fields = 9;
}

message EmbedField {
  string name = 1;
  string value = 2;
  bool inline = 3;
}

message SendEmbedRequest {
  string channel_id = 1;
  EmbedData embed = 2;
  string token = 3;
}

message SendEmbedResponse {
  bool success = 1;
  string message_id = 2;
  string error = 3;
}

message AddReactionRequest {
  string channel_id = 1;
  string message_id = 2;
  string emoji = 3;
  string token = 4;
}

message AddReactionResponse {
  bool success = 1;
  string error = 2;
}

message ManageRoleRequest {
  string guild_id = 1;
  string user_id = 2;
  string role_id = 3;
  string action = 4;  // "add" or "remove"
  string token = 5;
}

message ManageRoleResponse {
  bool success = 1;
  string error = 2;
}

message CreateWebhookRequest {
  string channel_id = 1;
  string name = 2;
  string token = 3;
}

message CreateWebhookResponse {
  bool success = 1;
  string webhook_url = 2;
  string error = 3;
}

message SendWebhookRequest {
  string webhook_url = 1;
  string content = 2;
  repeated EmbedData embeds = 3;
  string token = 4;
}

message SendWebhookResponse {
  bool success = 1;
  string error = 2;
}

message DeleteMessageRequest {
  string channel_id = 1;
  string message_id = 2;
  string token = 3;
}

message DeleteMessageResponse {
  bool success = 1;
  string error = 2;
}

message EditMessageRequest {
  string channel_id = 1;
  string message_id = 2;
  string content = 3;
  string token = 4;
}

message EditMessageResponse {
  bool success = 1;
  string error = 2;
}

message KickUserRequest {
  string guild_id = 1;
  string user_id = 2;
  string reason = 3;
  string token = 4;
}

message KickUserResponse {
  bool success = 1;
  string error = 2;
}

message BanUserRequest {
  string guild_id = 1;
  string user_id = 2;
  string reason = 3;
  int32 delete_message_days = 4;
  string token = 5;
}

message BanUserResponse {
  bool success = 1;
  string error = 2;
}

message TimeoutUserRequest {
  string guild_id = 1;
  string user_id = 2;
  int32 duration_seconds = 3;
  string reason = 4;
  string token = 5;
}

message TimeoutUserResponse {
  bool success = 1;
  string error = 2;
}
